compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/TransformUtils.kt

```diff
@@ -96,13 +98,20 @@ internal fun FirValueParameter.transformVarargTypeToArrayType() {
 }
 
 internal fun FirTypedDeclaration.transformTypeToArrayType() {
+    val returnTypeRef = this.returnTypeRef
+    require(returnTypeRef is FirResolvedTypeRef)
     val returnType = returnTypeRef.coneType
+
     transformReturnTypeRef(
         StoreType,
-        returnTypeRef.withReplacedConeType(
-            ConeKotlinTypeProjectionOut(returnType).createArrayType(),
-            FirFakeSourceElementKind.ArrayTypeFromVarargParameter
-        )
+        buildResolvedTypeRef {
+            source = returnTypeRef.source
+            type = ConeKotlinTypeProjectionOut(returnType).createArrayType()
+            annotations += returnTypeRef.annotations
+            delegatedTypeRef = returnTypeRef.delegatedTypeRef?.apply {
+                replaceSource(source?.fakeElement(FirFakeSourceElementKind.ArrayTypeFromVarargParameter))
+            }
```

Comment:
The original purpose of `delegatedTypeRef` was to leave reference to the previous type when a resolve type ref is built. E.g., type refs would be either implicit type ref, implicit built-ins, or user type ref, and when they're resolved, the generated resolved type ref points to the previous one via this reference. It would be useful, e.g., to distinguish user-specified type v.s. implicit type. But, such differentiation is implemented differently: actually, using the `source` instead, which means this `delegatedTypeRef` is not quite used at all.

So, here is one suggestion. When creating an array type from the original return type, we can set the original return type as delegated type ref:
```kt
transformReturnTypeRef(
    ...
    buildResolvedTypeRef(
        ...
        delegatedTypeRef = returnTypeRef.apply { replaceSource ... }
    }
}
```
Then, in the beginning of this util, we can bail out if the resolved return type has yet another resolved type as a delegated type ref, which means, this type has been transformed from a regular one to an array type. To be more accurate, we can double-check such delegated type ref has the fake source we set here.

Reply:
I like this approach :)

Reply:
Good approach

Reply:
Yeah, so hopefully, we can have a chain like this: array type -> vararg element type (-> user type ref).

Reply:
Thanks! Doing it this way seems to work correctly.